data_processing/eia.py

```diff
@@ -0,0 +1,2999 @@
+
+    if year not in allowed_years:
+
+        raise ValueError(
+            f"Not a valid year. Possibilities include: {*allowed_years,}."
+        )
+
+    headers = {
+        '2008': 7,
+        '2009': 7,
+        '2010': 7,
+        '2011': 5,
+        '2012': 5,
+        '2013': 5,
+        '2014': 5,
+        '2015': 5,
+        '2016': 5,
+        '2017': 5,
+        '2018': 5
+    }
+
+    netGen = {
+        '2008': 'NET GENERATION (megawatthours)',
+        '2009': 'NET GENERATION (megawatthours)',
+        '2010': 'NET GENERATION (megawatthours)',
+        '2011': 'Net Generation (Megawatthours)',
+        '2012': 'Net Generation (Megawatthours)',
+        '2013': 'Net Generation (Megawatthours)',
+        '2014': 'Net Generation (Megawatthours)',
+        '2015': 'Net Generation (Megawatthours)',
+        '2016': 'Net Generation (Megawatthours)',
+        '2017': 'Net Generation (Megawatthours)',
+        '2018': 'Net Generation (Megawatthours)'
+    }
+
+    generation = pd.read_csv((path + year + '.csv'),
+                             header=headers[year],
+                             usecols=['Operator Name',
+                                      'NERC Region',
+                                      'NAICS Code',
+                                      'Reported Prime Mover',
+                                      'Reported Fuel Type Code',
+                                      str(netGen[year])])
+
+    generation = generation.loc[generation['NAICS Code'] == 611]
+
+    generation['University'] = generation['Operator Name'].astype(
+        str) + ' ' + generation['NERC Region']
+
+    generation = generation.drop(
+        ['Operator Name', 'NERC Region'],
+        axis=1
+    )
+
+    university_keys = generation['University'].values.tolist()
+
+    final = dict()
+
+    fuel_type_list = list()
+
+    for uni in university_keys:
+
+        totalgen = 0
+
+        fuel_dict = dict()
+
+        iterable_dataframe = generation.loc[
+            generation['University'] == uni
+        ]
+
+        fuel_type_list = iterable_dataframe[
+            'Reported Fuel Type Code'
+        ].unique().tolist()
+
+        for fuel in fuel_type_list:
+
+            mover_dict = dict()
+
+            fuel_dataframe = iterable_dataframe.loc[
+                generation['University'] == uni
+            ].loc[
+                iterable_dataframe['Reported Fuel Type Code'] == fuel
+            ]
+
+            mover_list = fuel_dataframe[
+                'Reported Prime Mover'
+            ].values.tolist()
+
+            netgen_list = fuel_dataframe[
+                netGen[year]
+            ].values.tolist()
+
+            for i in range(len(mover_list)):
+
+                totalgen += netgen_list[i]
+
+                mover_dict.update({mover_list[i]: netgen_list[i]})
+
+            fuel_dict.update({fuel: mover_dict})
+
+        fuel_dict.update({'Total Generation': totalgen})
+
+        final.update({uni: fuel_dict})
+
+    final
```

Comment:
This can be done much more succinctly with Pandas. 

Not an exact suggestion because there are multiple Pandas solutions, I would start by using the ``groupby`` function. 

```python

#  get groups of items that all have the same naics code
naics = generation.groupby('NAICS Code')

#  get the universities
universities = naics.get_group(611)

#  regroup again by either fuel or something else
```

Here is a [good blog post](https://www.shanelynn.ie/summarising-aggregation-and-grouping-data-in-python-pandas/) that shows what I mean.